service/src/main/java/bio/terra/cbas/initialization/CloneRecoveryService.java

```diff
@@ -0,0 +1,96 @@
+package bio.terra.cbas.initialization;
+
+import bio.terra.cbas.config.CbasContextConfiguration;
+import bio.terra.cbas.dao.MethodDao;
+import bio.terra.cbas.dao.RunDao;
+import bio.terra.cbas.dao.RunSetDao;
+import bio.terra.cbas.models.Method;
+import bio.terra.cbas.models.Run;
+import bio.terra.cbas.models.RunSet;
+import bio.terra.common.db.WriteTransaction;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CloneRecoveryService {
+
+  private final Logger logger = LoggerFactory.getLogger(CloneRecoveryService.class);
+  private final RunDao runDao;
+  private final RunSetDao runSetDao;
+  private final MethodDao methodDao;
+  private final CbasContextConfiguration cbasContextConfig;
+
+  public CloneRecoveryService(
+      RunSetDao runSetDao,
+      RunDao runDao,
+      MethodDao methodDao,
+      CbasContextConfiguration cbasContextConfig) {
+    this.runSetDao = runSetDao;
+    this.runDao = runDao;
+    this.methodDao = methodDao;
+    this.cbasContextConfig = cbasContextConfig;
+  }
+
+  public void cloneRecovery() {
+    logger.info(
+        "Starting clone recovery (workspaceId: {}, workspaceCreatedDate: {})",
+        cbasContextConfig.getWorkspaceId(),
+        cbasContextConfig.getWorkspaceCreatedDate());
+
+    pruneCloneSourceWorkspaceHistory();
```

Comment:
I guess we then call other functions from inside the same class anyway, so this only moves the problem one call further down.

Hmm, this might be a use case for the `@Autowired CloneRecoveryService self` pattern that we saw in the spring intro

Reply:
hmm I see - good to know. I remember Doug saying that `@Autowired CloneRecoveryService self` can cause strange outcomes, so I may try to come up with some other way...